File: app/src/main/java/com/example/win/easy/view/main/MainActivityFragment.java
package com.example.win.easy.view.main;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;

import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;

import com.example.win.easy.R;
import com.example.win.easy.web.service.LoginService;
import com.qmuiteam.qmui.alpha.QMUIAlphaImageButton;
import com.qmuiteam.qmui.widget.QMUITopBar;
import com.qmuiteam.qmui.widget.grouplist.QMUICommonListItemView;
import com.qmuiteam.qmui.widget.grouplist.QMUIGroupListView;

import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * <p>点app图标进来后见到的主界面</p>
 */
public class MainActivityFragment extends Fragment {
    @BindView(R.id.listview) QMUIGroupListView itemGroup;
    @BindView(R.id.topbar) QMUITopBar qmuiTopBar;
    @BindString(R.string.textOnItemToAllSong) String textOnItemToAllSong;
    @BindString(R.string.textOnItemToAllSongList) String textOnItemToAllSongList;
    @BindString(R.string.mainActivityFragmentTopBarTitle) String topBarTitle;

    private LoginService loginService;

    public MainActivityFragment(LoginService loginService){
        this.loginService=loginService;
    }

    /**
     * <ol>
     *     <li>初始化顶栏（包括标题和按钮）</li>
     *     <li>初始化前往所有歌单的item</li>
     *     <li>...........所有歌曲的....</li>
     * </ol>
     */
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState){
        View thisView=inflater.inflate(R.layout.fragment_main_activity,container,false);
        ButterKnife.bind(this,thisView);

        setUpTopBar();

        QMUICommonListItemView itemToAllSong = initItemToAllSong();
        QMUICommonListItemView itemToAllSongList = initItemToAllSongList();
        showItems(itemToAllSong, itemToAllSongList);

        return thisView;
    }

    private void setUpTopBar(){
        qmuiTopBar.setTitle(topBarTitle);
        setUpButtonToLogin();
    }

    private void setUpButtonToLogin(){
        QMUIAlphaImageButton loginButton= qmuiTopBar.addLeftImageButton(R.drawable.ic_action_cloud, R.id.loginButtonId);
        loginButton.setOnClickListener(v->{
            if (!loginService.hasLogin())
                Navigation.findNavController(getView()).navigate(MainActivityFragmentDirections.actionMainActivityFragmentToLoginFragment());
        });
    }

    private QMUICommonListItemView initItemToAllSong(){
        QMUICommonListItemView itemView=newItem(textOnItemToAllSong, v -> Navigation.findNavController(getView()).navigate(MainActivityFragmentDirections.actionMainActivityFragmentToAllSongsFragment()));
        ImageButton buttonToAddLocalSong = initButtonToAddLocalSong();
        addButtonTo(buttonToAddLocalSong,itemView);
        return itemView;
    }

    private QMUICommonListItemView newItem(String text, View.OnClickListener onClickListener){
        QMUICommonListItemView itemView= itemGroup.createItemView(text);
        itemView.setAccessoryType(QMUICommonListItemView.ACCESSORY_TYPE_CUSTOM);
        itemView.setOnClickListener(onClickListener);
        return itemView;
    }

    private ImageButton initButtonToAddLocalSong(){
        return newImageButton(R.drawable.ic_action_name,v-> Navigation.findNavController(getView()).navigate(MainActivityFragmentDirections.actionMainActivityFragmentToPlaceholder()));
    }

    private ImageButton newImageButton(int imageResId, View.OnClickListener onClickListener){
        ImageButton imageButton=new ImageButton(getContext());
        imageButton.setImageResource(imageResId);
        imageButton.setOnClickListener(onClickListener);
        return imageButton;
    }

    private void addButtonTo(ImageButton imageButton,QMUICommonListItemView itemView){
        itemView.addAccessoryCustomView(imageButton);
    }

    private QMUICommonListItemView initItemToAllSongList(){
        QMUICommonListItemView itemView=newItem(textOnItemToAllSongList, v -> Navigation.findNavController(getView()).navigate(MainActivityFragmentDirections.actionMainActivityFragmentToAllSongListsFragment()));
        ImageButton buttonToCreateSongList = initButtonToCreateSongList();
        addButtonTo(buttonToCreateSongList,itemView);
        return itemView;
    }

    private ImageButton initButtonToCreateSongList(){
        return newImageButton(R.drawable.ic_action_name,v->Navigation.findNavController(getView()).navigate(MainActivityFragmentDirections.actionMainActivityFragmentToSongListCreationFragment()));
    }

    private void showItems(QMUICommonListItemView... itemViews){
        QMUIGroupListView.Section section= QMUIGroupListView.newSection(getContext());
        for (QMUICommonListItemView itemView:itemViews)
            section.addItemView(itemView,null);
        section.addTo(itemGroup);
    }
}

File: app/src/main/java/com/example/win/easy/view/main/MainActivityFragment.kt
package com.example.win.easy.view.main

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import androidx.fragment.app.Fragment
import androidx.navigation.Navigation
import butterknife.BindString
import butterknife.BindView
import butterknife.ButterKnife
import com.example.win.easy.R
import com.example.win.easy.network.LoginService
import com.qmuiteam.qmui.widget.QMUITopBar
import com.qmuiteam.qmui.widget.grouplist.QMUICommonListItemView
import com.qmuiteam.qmui.widget.grouplist.QMUIGroupListView

/**
 *
 * 点app图标进来后见到的主界面
 */
class MainActivityFragment(private val loginService: LoginService) : Fragment() {
    @BindView(R.id.listview)
    lateinit var itemGroup: QMUIGroupListView

    @BindView(R.id.topbar)
    lateinit var qmuiTopBar: QMUITopBar

    @BindString(R.string.textOnItemToAllSong)
    lateinit var textItemAllSong: String

    @BindString(R.string.textOnItemToAllSongList)
    lateinit var textItemAllSongList: String

    @BindString(R.string.mainActivityFragmentTopBarTitle)
    lateinit var topBarTitle: String

    /**
     *  主界面的初始化工作：
     *  1. 顶栏标题
     *  2. 登录按钮
     *  3. 前往所有歌曲/歌单的item
     *
     */
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.fragment_main_activity, container, false)
                    .also {
                        ButterKnife.bind(this, it)
                        setTopBar(topBarTitle)
                        setButtonLogin()
                        setItems(createItemAllSong(), createItemAllSongList())
                    }

    /**
     * 设置顶栏标题
     */
    private fun setTopBar(title: String) {
        qmuiTopBar.setTitle(title)
    }

    /**
     * 设置前往登录界面的按钮
     */
    private fun setButtonLogin(){
        qmuiTopBar.addLeftImageButton(R.drawable.ic_action_cloud, R.id.loginButtonId).run {
            setOnClickListener{
                synchronized(loginService) {
                    if (!loginService.hasLogin())
                        Navigation.findNavController(view!!)
                                .navigate(MainActivityFragmentDirections.actionMainActivityFragmentToLoginFragment())
                }
            }
        }
    }

    /**
     * 创建前往所有歌曲的item
     */
    private fun createItemAllSong() =
            newItem(textItemAllSong, View.OnClickListener {
                Navigation.findNavController(view!!)
                        .navigate(MainActivityFragmentDirections.actionMainActivityFragmentToAllSongsFragment())
            }).also {
                addButtonTo(createButtonAddSong(),it)
            }

    /**
     * 创建新的item
     */
    private fun newItem(text: String?, onClickListener: View.OnClickListener) =
            itemGroup.createItemView(text).apply {
                accessoryType = QMUICommonListItemView.ACCESSORY_TYPE_CUSTOM
                setOnClickListener(onClickListener)
            }

    /**
     * 创建添加歌曲的按钮
     */
    private fun createButtonAddSong()=
            newImageButton(R.drawable.ic_action_name, View.OnClickListener {
                Navigation.findNavController(view!!)
                        .navigate(MainActivityFragmentDirections.actionMainActivityFragmentToPlaceholder())
            })


    /**
     * 创建新的带图片的按钮
     */
    private fun newImageButton(imageResId: Int, onClickListener: View.OnClickListener)=
            ImageButton(context).apply{
                setImageResource(imageResId)
                setOnClickListener(onClickListener)
            }

    /**
     * 将按钮attach到item上
     */
    private fun addButtonTo(imageButton: ImageButton, itemView: QMUICommonListItemView) {
        itemView.addAccessoryCustomView(imageButton)
    }

    /**
     * 创建前往所有歌单的item
     */
    private fun createItemAllSongList() =
        newItem(textItemAllSongList,View.OnClickListener {
            Navigation.findNavController(view!!)
                    .navigate(MainActivityFragmentDirections.actionMainActivityFragmentToAllSongListsFragment())
        }).also {
            addButtonTo(createButtonAddSongList(),it)
        }

    /**
     * 创建添加歌单的按钮
     */
    private fun createButtonAddSongList()=
            newImageButton(
                    R.drawable.ic_action_name, View.OnClickListener {
                        Navigation.findNavController(view!!)
                                .navigate(MainActivityFragmentDirections.actionMainActivityFragmentToSongListCreationFragment())
            })

    /**
     * 设置并显示主界面的items
     */
    private fun setItems(vararg itemViews: QMUICommonListItemView) {
        QMUIGroupListView.newSection(context).run {
            for (itemView in itemViews)
                addItemView(itemView, null)
            addTo(itemGroup)
        }
    }

}